lib/internal/modules/cjs/loader.js

```diff
@@ -225,15 +226,41 @@ function readPackage(requestPath) {
   }
 }
 
-function tryPackage(requestPath, exts, isMain) {
-  var pkg = readPackage(requestPath);
+function tryPackage(requestPath, exts, isMain, originalPath) {
+  const pkg = readPackage(requestPath);
 
-  if (!pkg) return false;
+  if (!pkg) {
+    return tryExtensions(path.resolve(requestPath, 'index'), exts, isMain);
+  }
 
-  var filename = path.resolve(requestPath, pkg);
-  return tryFile(filename, isMain) ||
-         tryExtensions(filename, exts, isMain) ||
-         tryExtensions(path.resolve(filename, 'index'), exts, isMain);
+  const filename = path.resolve(requestPath, pkg);
+  let actual = tryFile(filename, isMain) ||
+    tryExtensions(filename, exts, isMain) ||
+    tryExtensions(path.resolve(filename, 'index'), exts, isMain);
+  if (actual === false) {
+    actual = tryExtensions(path.resolve(requestPath, 'index'), exts, isMain);
+    if (!actual) {
+      // eslint-disable-next-line no-restricted-syntax
+      const err = new Error(
+        `Cannot find module '${filename}'. ` +
+        'Please verify that the package.json has a valid "main" entry'
+      );
+      err.code = 'MODULE_NOT_FOUND';
+      err.path = path.resolve(requestPath, 'package.json');
+      err.requestPath = originalPath;
+      // TODO(BridgeAR): Add the requireStack as well.
+      throw err;
+    } else if (pendingDeprecation) {
+      const packageJSON = path.resolve(requestPath, 'package.json');
+      process.emitWarning(
+        `The 'package.json' (${packageJSON}) has an invalid 'main' entry ` +
+          `(${pkg}). Please either fix that or report it to the module author.`,
```

Comment:
`entry` -> `field`.

I think writing `'package.json'` and then putting a path to the `package.json` is a bit of duplication. Maybe it could be worded like:

```js
`Invalid 'main' field in '${jsonPath}' of '${fieldValue}'. ` +
`Please either fix that or report it to the module author`
```

Also note no period on the second sentence.